### DISCUSSION TITLE
Does Microsoft.Extensions.Http.Resilience works for Blazor project?

### DISCUSSION THREAD

**Discussion by aDisplayName** [dotnet/extensions#5234] [Q&A]
The way HttpClient is configured in Blazor WASM project is via the use of `IServiceCollection.AddScoped<T>()`
```csharp
builder.Services
    .AddScoped(sp => new HttpClient 
        { BaseAddress = new Uri(builder.HostEnvironment.BaseAddress) })
```

The extension method of `IHttpClientBuilder.AddStandardResilienceHandler` is an extension on `IHttpClientBuilder`. So my question is can `Microsoft.Extensions.Http.Resilience` be used for Blazor WASM project?

I tried to disable the `AddScoped` in setup, and instead of using 
```csharp
builder.Services.AddHttpClient<MyService>(x => x.BaseAddress = new Uri(builder.HostEnvironment.BaseAddress))
    .AddStandardResilienceHandler();
```
and it seems it doesn't work at all, an exception is thrown:
```
System.InvalidOperationException An invalid request URI was provided. Either the request URI must be an absolute URI or BaseAddress must be set
```

Would be great if the resilience library be used for Blazor as well.

@martintmk @danroth27

**Comment by danroth27** [comment]
That error basically says the HttpClient is being asked to make a request to a relative URI without a specified base address. What's the stack trace for that exception? Can you confirm whether the used HttpClient has a valid base address or not?

**Reply by aDisplayName** [reply]
Hi, @danroth27 ,

So the original code in Program.cs in my Blazor WASM is

```csharp
using Microsoft.AspNetCore.Components.Web;
using Microsoft.AspNetCore.Components.WebAssembly.Hosting;
using Microsoft.FluentUI.AspNetCore.Components;
using Slb.Everest.Wbi.Sealy.Portal.Web.Client.Services;

var builder = WebAssemblyHostBuilder.CreateDefault(args);


builder.Services
    .AddScoped(sp => new HttpClient { BaseAddress = new Uri(builder.HostEnvironment.BaseAddress) })
    .AddScoped<MyService>();

builder.Services.AddFluentUIComponents();
builder.Services.AddTelerikBlazor();
builder.RootComponents.Add<HeadOutlet>("head::after");

await builder.Build().RunAsync();
```

And it works without any problem.

In order to use the `IHttpClientBuilder.AddStandardResilienceHandler`, I've added Microsoft.Extensions.Http.Resilience nuget package, and changed the program.cs into
```csharp
using Microsoft.AspNetCore.Components.Web;
using Microsoft.AspNetCore.Components.WebAssembly.Hosting;
using Microsoft.FluentUI.AspNetCore.Components;
using Slb.Everest.Wbi.Sealy.Portal.Web.Client.Services;

var builder = WebAssemblyHostBuilder.CreateDefault(args);

builder.Services.AddHttpClient<ProjectGatewayService>(x => x.BaseAddress = new Uri(builder.HostEnvironment.BaseAddress))
    .AddStandardResilienceHandler();

builder.Services
    // Replaced with .AddHttpClient<> above
    // .AddScoped(sp => new HttpClient { BaseAddress = new Uri(builder.HostEnvironment.BaseAddress) })
    .AddScoped<MyService>();


builder.Services.AddFluentUIComponents();
builder.Services.AddTelerikBlazor();
builder.RootComponents.Add<HeadOutlet>("head::after");

await builder.Build().RunAsync();
```
And the exception thrown when calling the inject the httpclient when calling with a relative path. Apparently the BaseAddress was not being set.

I guess the actual question is, does Blazor app support using `IHttpClientBuilder AddHttpClient<TClient>(this IServiceCollection services, Action<HttpClient> configureClient)` extenstion method to inject HttpClient.
